entoas/edges.go

```diff
@@ -0,0 +1,117 @@
+package entoas
+
+import (
+	"fmt"
+
+	"entgo.io/contrib/entoas/serialization"
+	"entgo.io/ent/entc/gen"
+)
+
+const _maxDepth = 25
+
+type (
+	// Edge wraps a gen.Edge and denotes an edge to be returned in an operation response. It recursively defines
+	// edges to load on the gen.Type the wrapped edge is pointing at.
+	Edge struct {
+		*gen.Edge
+		Edges Edges
+	}
+	Edges []*Edge
+	// walk is a node sequence in the schema graph. Used to keep track when creating an Edge.
+	walk []string
+)
+
+// EdgeTree returns the Edges to include on a type for the given serialization groups.
+func EdgeTree(n *gen.Type, gs serialization.Groups) (Edges, error) { return edgeTree(n, walk{}, gs) }
+
+// Flatten returns a list of all gen.Types present in the tree.
+func (es Edges) Flatten() []*gen.Edge {
```

Comment:
Any reason `EdgeTree` does not call flatten internally? I guess it won't be a tree anymore, but it looks like this separation is not needed.

Reply:
This is to have an possible implementation of the spec be able to construct the query to eager load edges. For that the tree needs to be intact.